test/gtest/tensor_api.cpp

```diff
@@ -0,0 +1,437 @@
+                                         dims,
+                                         dims + 1,
+                                         dims + 2,
+                                         dims + 3,
+                                         strides,
+                                         strides + 1,
+                                         strides + 2,
+                                         strides + 3);
+    if(status != miopenStatusSuccess)
+        return TestStatus::Failed;
+
+    if(params.dataType != dataType || !CompareLengths(params.dimsA, dims, 4) ||
+       (params.use_strides && !CompareLengths(params.stridesA, strides, 4)))
+        return TestStatus::Failed;
+
+    return TestStatus::Passed;
+}
+
+TestStatus GetTensorDescriptor(miopenTensorDescriptor_t tensorDesc, const TensorParams& params)
+{
+    if(params.dimsA == nullptr || (params.stridesA == nullptr && params.use_strides))
+        return TestStatus::Failed; // internal error
+
+    miopenStatus_t status;
+    int size;
+
+    status = miopenGetTensorDescriptorSize(tensorDesc, &size);
+    if(status != miopenStatusSuccess || size < 0 || size != params.nbDims)
+        return TestStatus::Failed;
+
+    miopenDataType_t dataType;
+    std::vector<int> dims(size);
+    std::vector<int> strides(size);
+
+    status = miopenGetTensorDescriptor(tensorDesc, &dataType, dims.data(), strides.data());
+    if(status != miopenStatusSuccess)
+        return TestStatus::Failed;
+
+    if(params.dataType != dataType || !CompareLengths(params.dimsA, dims, size) ||
+       (params.use_strides && !CompareLengths(params.stridesA, strides, size)))
+        return TestStatus::Failed;
+
+    return TestStatus::Passed;
+}
+
+const auto get_tensor_descr_funcs = {Get4dTensorDescriptor, GetTensorDescriptor};
+
+// Generate test data
+void GenerateValidTestConfigs(std::vector<TestConfig>& configs, bool use_strides)
+{
+    const auto datatypes = {miopenHalf, miopenDouble};
+    const auto layouts   = {miopenTensorNCHW, miopenTensorNCDHW};
+    static int dims[]    = {4, 4, 16, 9, 16};
+    static int strides[] = {9216, 2304, 144, 16, 1};
+    static_assert(sizeof(dims) == sizeof(strides));
+    const auto max_ndims = sizeof(dims) / sizeof(dims[0]);
+
+    for(auto datatype : datatypes)
+    {
+        for(auto layout : layouts)
+        {
+            for(int ndims = 1; ndims <= max_ndims; ndims++)
+            {
+                const TestConfig config = {false,
+                                           {datatype,
+                                            layout,
+                                            ndims,
+                                            dims + (max_ndims - ndims),
+                                            use_strides ? (strides + (max_ndims - ndims)) : nullptr,
+                                            use_strides},
+                                           true};
+                configs.push_back(config);
```

Comment:
This would be more performant and less memory consuming with a visitor pattern: rather than storing them to a vector this could take a `std::function<void(const TestConfig&)>` and just call it to consume the result. So from  `TEST_F(TestTensorApi, SetTensor)` it could be called like this
```C++
    RunValidTestConfigs<true>([](auto&& config){ RunTest(config)); });
    RunValidTestConfigs<false>([](auto&& config){ RunTest(config)); });
```
So at any time only one config is actually in memory.

Reply:
@averinevg could you follow up on this one?

Reply:
> @averinevg could you follow up on this one?

@junliume Sure